File: scripts/normalize-transcribed.js
const { basename } = require('path');
const { existsSync, readFileSync } = require('fs');;
const { print, printError } = require('./_printers.js')


function printHelp() {
	print(`Usage ${basename(process.argv[1])} WORD-LIST.txt`);
	print('Normalizes the frequencies in a dictionary with transcriptions.');
}



function validateInput() {
	if (process.argv.length < 3) {
		printHelp();
		process.exit(1);
	}

	if (!existsSync(process.argv[2])) {
		printError(`Failure! Could not find word list file "${process.argv[2]}".`);
		process.exit(2);
	}

	return {
		fileName: process.argv[2]
	};
}


function printWords(wordList) {
	if (Array.isArray(wordList)) {
		wordList.forEach(w => print(
			w.number !== null ? `${w.chinese}\t${w.latin}\t${w.number}` : `${w.chinese}\t${w.latin}`
		));
	}
}


const { fileName } = validateInput();

const data = readFileSync(fileName, 'utf8');
const lines = data.trim().split('\n');

// Parse the data into an array of objects
let entries = lines.map(line => {
	const parts = line.split('\t');
	return {
		original: line,
		chinese: parts[0],
		latin: parts[1],
		number: parts[2] ? parseInt(parts[2], 10) : null
	};
});

// Group entries by the Latin character sequence
const groups = {};
entries.forEach(entry => {
	if (!groups[entry.latin]) {
		groups[entry.latin] = [];
	}
	groups[entry.latin].push(entry);
});

// Process each group: sort by number (descending) and reassign ordinal numbers
let sortedEntries = [];
for (const key in groups) {
	let group = groups[key];

	// Separate entries with and without numbers
	let withNumbers = group.filter(e => e.number !== null);
	let withoutNumbers = group.filter(e => e.number === null);

	// Sort by number in descending order
	withNumbers.sort((a, b) => b.number - a.number);

	// Assign ordinal rankings
	for (let i = 0; i < withNumbers.length; i++) {
		withNumbers[i].number = (withNumbers.length - i).toString();
	}

	// Preserve original order for entries without numbers
	sortedEntries.push(...withNumbers, ...withoutNumbers);
}

printWords(sortedEntries);

File: scripts/normalize-transcribed.py
import sys
import os
import argparse
import yaml
from collections import defaultdict

def print_error(message):
    print(message, file=sys.stderr)

# ---------- Argument Parsing ----------
def parse_args():
    parser = argparse.ArgumentParser(
        description="Normalizes the frequencies in a dictionary with transcriptions."
    )
    parser.add_argument(
        "word_list",
        help="Path to the word list file (e.g., WORD-LIST.txt)"
    )
    parser.add_argument(
        "layout_yaml",
        help="Path to the YAML file containing the layout definitions (UTF-8 encoded)"
    )
    return parser.parse_args()

# ---------- File Validation ----------
def validate_file(file_path):
    if not os.path.isfile(file_path):
        print_error(f'Failure! Could not find file "{file_path}".')
        sys.exit(2)

# ---------- YAML Layout Loading ----------
def load_layout(yaml_path):
    with open(yaml_path, encoding='utf-8') as f:
        data = yaml.safe_load(f)

    if "layout" not in data or not isinstance(data["layout"], list):
        print_error("Error: YAML file must contain a 'layout' key with a list of lists.")
        sys.exit(4)

    layout_dict = {}
    seen = set()

    for index, group in enumerate(data["layout"]):
        if not isinstance(group, list) or len(group) == 0:
            print_error(f"Error: Layout entry {index} must be a non-empty list of strings.")
            sys.exit(4)
        for symbol in group:
            if symbol in seen:
                print_error(f"Error: Duplicate symbol '{symbol}' found in layout. Aborting.")
                sys.exit(4)
            seen.add(symbol)
            layout_dict[symbol] = index

    return layout_dict

# ---------- Word List Loading ----------
def load_entries(file_path):
    with open(file_path, encoding='utf-8') as f:
        lines = [line.strip() for line in f if line.strip()]

    entries = []
    for line_num, line in enumerate(lines, start=1):
        parts = line.split('\t')
        if len(parts) < 2:
            print_error(f"Malformed line {line_num}: '{line}' (expected at least 2 tab-separated fields)")
            sys.exit(3)

        native, latin = parts[:2]
        number = None
        if len(parts) > 2:
            try:
                number = int(parts[2])
            except ValueError:
                print_error(f"Malformed line {line_num}: '{line}' (third field must be an integer if present)")
                sys.exit(3)

        entries.append({'native': native, 'latin': latin, 'number': number})

    return entries

# ---------- Grouping by Layout Index Pattern ----------
def group_entries(entries, layout_dict):
    groups = defaultdict(list)

    # Sort symbols by length (descending) for multi-letter matching (e.g., 'Zh' before 'Z')
    sorted_symbols = sorted(layout_dict.keys(), key=len, reverse=True)

    for entry in entries:
        latin = entry['latin']
        i = 0
        index_seq = []

        while i < len(latin):
            matched = False
            for symbol in sorted_symbols:
                if latin.startswith(symbol, i):
                    index_seq.append(str(layout_dict[symbol]))
                    i += len(symbol)
                    matched = True
                    break
            if not matched:
                print_error(f"Error: Unknown symbol in Latin string '{latin}' near '{latin[i:]}'")
                sys.exit(5)

        key = ''.join(index_seq)
        groups[key].append(entry)

    return groups

# ---------- Frequency Normalization ----------
def normalize_frequencies(groups):
    sorted_entries = []
    for group in groups.values():
        with_numbers = [e for e in group if e['number'] is not None]
        without_numbers = [e for e in group if e['number'] is None]

        with_numbers.sort(key=lambda e: e['number'], reverse=True)

        for rank, entry in enumerate(with_numbers, start=1):
            entry['number'] = str(len(with_numbers) - rank + 1)

        sorted_entries.extend(with_numbers)
        sorted_entries.extend(without_numbers)

    return sorted_entries

# ---------- Output ----------
def print_entries(entries):
    for e in entries:
        parts = [e['native'], e['latin']]
        if e['number'] is not None:
            parts.append(str(e['number']))
        print('\t'.join(parts))

# ---------- Main ----------
def main():
    args = parse_args()
    validate_file(args.word_list)
    validate_file(args.layout_yaml)

    layout_dict = load_layout(args.layout_yaml)
    entries = load_entries(args.word_list)
    groups = group_entries(entries, layout_dict)
    sorted_entries = normalize_frequencies(groups)
    print_entries(sorted_entries)

if __name__ == "__main__":
    main()
